src/main/java/io/cdap/plugin/ga360/source/common/requests/ReportsRequestFactory.java

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright © 2019 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package io.cdap.plugin.ga360.source.common.requests;
+
+import com.google.api.services.analyticsreporting.v4.model.DateRange;
+import com.google.api.services.analyticsreporting.v4.model.Dimension;
+import com.google.api.services.analyticsreporting.v4.model.GetReportsRequest;
+import com.google.api.services.analyticsreporting.v4.model.GetReportsResponse;
+import com.google.api.services.analyticsreporting.v4.model.Metric;
+import com.google.api.services.analyticsreporting.v4.model.Report;
+import com.google.api.services.analyticsreporting.v4.model.ReportRequest;
+import io.cdap.plugin.ga360.source.batch.GoogleAnalyticsBatchSourceConfig;
+import io.cdap.plugin.ga360.source.common.AnalyticsReportingInitializer;
+import io.cdap.plugin.ga360.source.common.BaseSourceConfig;
+
+import java.io.IOException;
+import java.security.GeneralSecurityException;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * Creates batch request based on source configuration.
+ */
+public class ReportsRequestFactory {
+
+  /**
+   * Fetches page of reports.
+   */
+  public static Iterator<Report> requestNextReports(BaseSourceConfig config, String nextPageToken) {
+    try {
+      ReportRequest reportRequest = ReportsRequestFactory.createRequest(config);
+      reportRequest.setPageToken(nextPageToken);
+
+      //for batch source only
+      if (config instanceof GoogleAnalyticsBatchSourceConfig) {
```

Comment:
The utility class should not have this kind of check, you can change the parameter to just accept the `ReportRequest` and set this beforehand.

Reply:
removed